File: hmh-backend/app/schemas/fuel_management.py
"""Pydantic v2 contracts for Fuel Management."""

import uuid
from datetime import date, datetime, timezone
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


class OrmModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class FuelTypeRead(OrmModel):
    id: uuid.UUID
    code: str
    name: str
    is_active: bool


class FuelStorageCreate(BaseModel):
    site_id: Optional[uuid.UUID] = None
    fuel_type_id: uuid.UUID
    name: str = Field(min_length=2, max_length=160)
    location_type: str = "TANK"
    capacity_litres: Optional[float] = Field(default=None, gt=0)
    low_stock_threshold_litres: Optional[float] = Field(default=None, ge=0)
    notes: Optional[str] = None
    opening_stock_litres: float = Field(default=0, ge=0)


class FuelStorageRead(OrmModel):
    id: uuid.UUID
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID]
    fuel_type_id: uuid.UUID
    name: str
    location_type: str
    capacity_litres: Optional[float]
    low_stock_threshold_litres: Optional[float]
    is_active: bool
    notes: Optional[str]
    calculated_balance_litres: float = 0
    cutover_confirmed_at: Optional[datetime] = None
    cutover_confirmed_by: Optional[uuid.UUID] = None


class FuelOrderCreate(BaseModel):
    site_id: Optional[uuid.UUID] = None
    fuel_type_id: uuid.UUID
    supplier_id: Optional[uuid.UUID] = None
    storage_location_id: Optional[uuid.UUID] = None
    request_date: date = Field(default_factory=date.today)
    requested_litres: float = Field(gt=0)
    expected_delivery_date: Optional[date] = None
    delivery_location: str = Field(min_length=2, max_length=300)
    purpose: Optional[str] = None
    intended_use: Optional[str] = Field(default=None, max_length=80)
    destination_type: Optional[Literal[
        "VEHICLE", "SITE_STORAGE", "GENERATOR", "PLANT", "OTHER_EQUIPMENT"
    ]] = None
    vehicle_id: Optional[uuid.UUID] = None
    equipment_reference: Optional[str] = Field(default=None, max_length=200)
    notes: Optional[str] = None
    submit_now: bool = False


class FuelOrderUpdate(BaseModel):
    supplier_id: Optional[uuid.UUID] = None
    storage_location_id: Optional[uuid.UUID] = None
    expected_delivery_date: Optional[date] = None
    delivery_location: Optional[str] = Field(default=None, min_length=2, max_length=300)
    requested_litres: Optional[float] = Field(default=None, gt=0)
    purpose: Optional[str] = None
    supplier_reference: Optional[str] = None
    purchase_order_reference: Optional[str] = None


class FuelTransition(BaseModel):
    reason: Optional[str] = None
    override_reason: Optional[str] = None
    supplier_reference: Optional[str] = None
    purchase_order_reference: Optional[str] = None


class FuelOrderHistoryRead(OrmModel):
    id: uuid.UUID
    from_status: Optional[str]
    to_status: str
    actor_id: uuid.UUID
    actor_name: Optional[str] = None
    reason: Optional[str]
    created_at: datetime


class FuelOrderRead(OrmModel):
    id: uuid.UUID
    order_number: str
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID]
    fuel_type_id: uuid.UUID
    supplier_id: Optional[uuid.UUID]
    storage_location_id: Optional[uuid.UUID]
    requested_by: uuid.UUID
    request_date: date
    requested_litres: float
    expected_delivery_date: Optional[date]
    delivery_location: str
    purpose: Optional[str]
    intended_use: Optional[str] = None
    destination_type: Optional[str] = None
    vehicle_id: Optional[uuid.UUID] = None
    equipment_reference: Optional[str] = None
    notes: Optional[str] = None
    status: str
    approved_by: Optional[uuid.UUID]
    approved_at: Optional[datetime]
    rejection_reason: Optional[str]
    supplier_reference: Optional[str]
    purchase_order_reference: Optional[str]
    submitted_at: Optional[datetime] = None
    feasibility_status: str = "NOT_EVALUATED"
    feasibility_message: Optional[str] = None
    estimated_remaining_litres: Optional[float] = None
    feasibility_override_reason: Optional[str] = None
    feasibility_override_by: Optional[uuid.UUID] = None
    feasibility_override_at: Optional[datetime] = None
    requester_name: Optional[str] = None
    next_approver: Optional[str] = None
    history: list[FuelOrderHistoryRead] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    delivered_litres: float = 0


class FuelDeliveryCreate(BaseModel):
    delivered_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    delivered_litres: float = Field(gt=0)
    confirmed_litres: Optional[float] = Field(default=None, gt=0)
    delivery_note_number: str = Field(min_length=1, max_length=120)
    supplier_id: Optional[uuid.UUID] = None
    storage_location_id: Optional[uuid.UUID] = None
    opening_reading: Optional[float] = Field(default=None, ge=0)
    closing_reading: Optional[float] = Field(default=None, ge=0)
    tanker_registration: Optional[str] = None
    driver_details: Optional[str] = None
    notes: Optional[str] = None
    allow_excess: bool = False
    excess_reason: Optional[str] = None

    @model_validator(mode="after")
    def readings_are_complete(self):
        if (self.opening_reading is None) != (self.closing_reading is None):
            raise ValueError("opening and closing readings must be supplied together")
        if self.opening_reading is not None and self.closing_reading < self.opening_reading:
            raise ValueError("closing reading cannot be lower than opening reading")
        if self.allow_excess and not self.excess_reason:
            raise ValueError("excess_reason is required for an excess override")
        return self


class FuelDeliveryRead(OrmModel):
    id: uuid.UUID
    order_id: Optional[uuid.UUID]
    procurement_delivery_item_id: Optional[uuid.UUID] = None
    project_id: Optional[uuid.UUID]
    site_id: uuid.UUID
    supplier_id: Optional[uuid.UUID]
    fuel_type_id: Optional[uuid.UUID]
    storage_location_id: Optional[uuid.UUID]
    delivery_date: date
    delivered_at: Optional[datetime]
    delivery_note_number: Optional[str]
    litres_delivered: float
    calculated_received_litres: Optional[float]
    confirmed_litres: Optional[float]
    variance_litres: Optional[float]
    supplier_variance_litres: Optional[float] = None
    meter_variance_litres: Optional[float] = None
    is_manual_emergency: bool = False
    emergency_reason: Optional[str] = None
    tanker_registration: Optional[str]
    driver_details: Optional[str]
    verification_status: str
    excess_override: bool
    notes: Optional[str]
    created_at: datetime


class FuelDeliveryFromProcurementCreate(BaseModel):
    """Hand-off from a real procurement DeliveryItem into the Fuel Control
    layer (Phase 5). litres_delivered is taken from the DeliveryItem itself
    (the office's documented/supplier quantity) — only the Fuel-side
    confirmation and optional meter readings are captured here."""
    delivery_item_id: uuid.UUID
    storage_location_id: uuid.UUID
    confirmed_litres: float = Field(gt=0)
    delivered_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    opening_reading: Optional[float] = Field(default=None, ge=0)
    closing_reading: Optional[float] = Field(default=None, ge=0)
    tanker_registration: Optional[str] = None
    driver_details: Optional[str] = None
    notes: Optional[str] = None

    @model_validator(mode="after")
    def readings_are_complete(self):
        if (self.opening_reading is None) != (self.closing_reading is None):
            raise ValueError("opening and closing readings must be supplied together")
        if self.opening_reading is not None and self.closing_reading < self.opening_reading:
            raise ValueError("closing reading cannot be lower than opening reading")
        return self


class FuelManualEmergencyDeliveryCreate(BaseModel):
    """Emergency receipt with no procurement chain behind it at all (e.g. a
    cash fuel purchase before a Material Request could be raised). Gated on
    fuel.admin at the API layer; reason is mandatory, not optional notes."""
    storage_location_id: uuid.UUID
    delivered_litres: float = Field(gt=0)
    confirmed_litres: Optional[float] = Field(default=None, gt=0)
    reason: str = Field(min_length=3)
    delivered_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    delivery_note_number: Optional[str] = None
    supplier_id: Optional[uuid.UUID] = None
    opening_reading: Optional[float] = Field(default=None, ge=0)
    closing_reading: Optional[float] = Field(default=None, ge=0)
    tanker_registration: Optional[str] = None
    driver_details: Optional[str] = None
    notes: Optional[str] = None

    @model_validator(mode="after")
    def readings_are_complete(self):
        if (self.opening_reading is None) != (self.closing_reading is None):
            raise ValueError("opening and closing readings must be supplied together")
        if self.opening_reading is not None and self.closing_reading < self.opening_reading:
            raise ValueError("closing reading cannot be lower than opening reading")
        return self


class FuelIssueCreate(BaseModel):
    site_id: Optional[uuid.UUID] = None
    storage_location_id: uuid.UUID
    fuel_type_id: uuid.UUID
    vehicle_id: Optional[uuid.UUID] = None
    destination_type: Literal["VEHICLE", "PLANT", "GENERATOR", "STORAGE_TANK", "OTHER_EQUIPMENT"]
    equipment_reference: Optional[str] = None
    issued_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    litres: float = Field(gt=0)
    odometer_reading: Optional[float] = Field(default=None, ge=0)
    hour_meter_reading: Optional[float] = Field(default=None, ge=0)
    received_by: Optional[str] = None
    purpose: Optional[str] = None
    evidence_url: Optional[str] = None
    notes: Optional[str] = None
    reading_source: Literal["MANUAL", "PHOTOGRAPH_VERIFIED", "TRACKER_VERIFIED", "MANAGER_OVERRIDDEN"] = "MANUAL"
    evidence_override_reason: Optional[str] = None
    feasibility_override_reason: Optional[str] = None


class FuelIssueRead(OrmModel):
    id: uuid.UUID
    issue_number: str
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID]
    storage_location_id: uuid.UUID
    fuel_type_id: uuid.UUID
    vehicle_id: Optional[uuid.UUID]
    destination_type: str
    equipment_reference: Optional[str]
    issued_at: datetime
    litres: float
    odometer_reading: Optional[float]
    hour_meter_reading: Optional[float]
    received_by: Optional[str]
    purpose: Optional[str]
    evidence_url: Optional[str]
    distance_since_previous_km: Optional[float]
    litres_per_100km: Optional[float]
    operating_hours_since_previous: Optional[float]
    litres_per_hour: Optional[float]
    anomaly_flag: bool
    anomaly_reason: Optional[str]
    reading_source: str = "MANUAL"
    tracker_provider: Optional[str] = None
    tracker_reading_at: Optional[datetime] = None
    estimated_remaining_litres: Optional[float] = None
    feasibility_status: str = "NOT_EVALUATED"
    evidence_override_reason: Optional[str] = None
    evidence_override_by: Optional[uuid.UUID] = None
    evidence_override_at: Optional[datetime] = None
    feasibility_override_reason: Optional[str] = None
    feasibility_override_by: Optional[uuid.UUID] = None
    feasibility_override_at: Optional[datetime] = None
    evidence: list[dict] = Field(default_factory=list)
    is_reversed: bool
    reversal_reason: Optional[str]
    created_at: datetime


class FuelAdjustmentCreate(BaseModel):
    storage_location_id: uuid.UUID
    adjustment_type: str
    litres_delta: float
    reason: str = Field(min_length=3)
    reference_reconciliation_id: Optional[uuid.UUID] = None


class FuelAdjustmentRead(OrmModel):
    id: uuid.UUID
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID]
    storage_location_id: uuid.UUID
    fuel_type_id: uuid.UUID
    adjustment_type: str
    litres_delta: float
    reason: str
    authorised_by: uuid.UUID
    created_at: datetime


class FuelReconciliationCreate(BaseModel):
    storage_location_id: uuid.UUID
    reconciliation_date: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    physical_balance_litres: float = Field(ge=0)
    explanation: str = Field(min_length=3)


class FuelReconciliationRead(OrmModel):
    id: uuid.UUID
    reconciliation_number: str
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID]
    storage_location_id: uuid.UUID
    fuel_type_id: uuid.UUID
    reconciliation_date: datetime
    calculated_balance_litres: float
    physical_balance_litres: float
    variance_litres: float
    variance_pct: Optional[float]
    explanation: str
    status: str
    requires_approval: bool
    reconciled_by: uuid.UUID
    approved_by: Optional[uuid.UUID]
    approval_notes: Optional[str]
    created_at: datetime


class FuelEquipmentProfileCreate(BaseModel):
    site_id: Optional[uuid.UUID] = None
    equipment_reference: str = Field(min_length=1, max_length=200)
    destination_type: str
    expected_litres_per_hour: Optional[float] = Field(default=None, gt=0)
    tolerance_pct: float = Field(default=20, ge=0, le=500)
    tank_capacity_litres: Optional[float] = Field(default=None, gt=0)
    minimum_issue_interval_hours: float = Field(default=0, ge=0)
    hour_meter_required: bool = True
    override_required: bool = False


class FuelEquipmentProfileRead(OrmModel):
    id: uuid.UUID
    project_id: uuid.UUID
    site_id: Optional[uuid.UUID]
    equipment_reference: str
    destination_type: str
    expected_litres_per_hour: Optional[float]
    tolerance_pct: float
    tank_capacity_litres: Optional[float]
    minimum_issue_interval_hours: float
    hour_meter_required: bool
    override_required: bool
    is_active: bool


class FuelEmailLogRead(OrmModel):
    id: uuid.UUID
    order_id: Optional[uuid.UUID]
    delivery_id: Optional[uuid.UUID]
    event_type: str
    recipient_email: str
    subject: str
    status: str
    attempt_count: int
    last_attempt_at: Optional[datetime]
    next_attempt_at: Optional[datetime]
    error_message: Optional[str]
    created_at: datetime

File: hmh-frontend/src/api/fuelManagement.ts
import client from "./client";

export interface FuelTypeDefinition { id: string; code: string; name: string; is_active: boolean }
export interface FuelStorage { id: string; project_id: string; site_id: string | null; fuel_type_id: string; name: string; location_type: string; capacity_litres: number | null; low_stock_threshold_litres: number | null; calculated_balance_litres: number; notes: string | null }
export interface FuelEquipmentProfile { id: string; project_id: string; site_id: string | null; equipment_reference: string; destination_type: string; is_active: boolean }
export interface FuelOrder { id: string; order_number: string; project_id: string; site_id: string | null; fuel_type_id: string; supplier_id: string | null; storage_location_id: string | null; requested_by: string; requester_name?: string | null; request_date: string; requested_litres: number; expected_delivery_date: string | null; delivery_location: string; purpose: string | null; intended_use?: string | null; destination_type?: string | null; vehicle_id?: string | null; equipment_reference?: string | null; notes?: string | null; status: string; supplier_reference: string | null; purchase_order_reference: string | null; delivered_litres: number; submitted_at?: string | null; next_approver?: string | null; feasibility_status?: string; feasibility_message?: string | null; estimated_remaining_litres?: number | null; history?: Array<{ id: string; from_status: string | null; to_status: string; actor_name: string | null; reason: string | null; created_at: string }>; created_at: string }
export interface FuelDelivery { id: string; order_id: string | null; procurement_delivery_item_id: string | null; project_id: string; site_id: string; storage_location_id: string | null; delivery_note_number: string | null; delivered_at: string | null; litres_delivered: number; confirmed_litres: number | null; variance_litres: number | null; supplier_variance_litres: number | null; meter_variance_litres: number | null; is_manual_emergency: boolean; emergency_reason: string | null; verification_status: string; excess_override: boolean; notes: string | null; created_at: string }
export interface FuelPendingConfirmation { delivery_item_id: string; description: string; quantity_received: number; unit: string | null; delivery_id: string; delivery_number: string | null; delivery_date: string; supplier_delivery_note_number: string | null; po_id: string; po_number: string; supplier_id: string }
export interface FuelIssue { id: string; issue_number: string; project_id: string; storage_location_id: string; fuel_type_id: string; vehicle_id: string | null; destination_type: string; equipment_reference: string | null; issued_at: string; litres: number; odometer_reading: number | null; hour_meter_reading: number | null; litres_per_100km: number | null; litres_per_hour: number | null; anomaly_flag: boolean; anomaly_reason: string | null; feasibility_status: string; reading_source: string; estimated_remaining_litres: number | null; evidence: Array<{ type: string; attachment_id: string }>; is_reversed: boolean; unit_cost: number | null; total_cost: number | null }
export interface FuelReconciliation { id: string; reconciliation_number: string; storage_location_id: string; reconciliation_date: string; calculated_balance_litres: number; physical_balance_litres: number; variance_litres: number; variance_pct: number | null; explanation: string; status: string; requires_approval: boolean }
export interface FuelDashboard { litres_requested: number; litres_approved: number; litres_ordered: number; litres_delivered: number; litres_issued: number; current_calculated_stock: number; outstanding_orders: number; overdue_deliveries: number; anomalies_for_review: number; storage_locations: { id: string; name: string; balance_litres: number }[] }

const data = <T>(promise: Promise<{ data: { data: T } }>) => promise.then((r) => r.data.data);

export const fuelManagementApi = {
  fuelTypes: () => data<FuelTypeDefinition[]>(client.get("/fuel-management/fuel-types")),
  dashboard: (projectId: string) => data<FuelDashboard>(client.get(`/projects/${projectId}/fuel-management/dashboard`)),
  storage: (projectId: string) => data<FuelStorage[]>(client.get(`/projects/${projectId}/fuel-management/storage`)),
  equipmentProfiles: (projectId: string) => data<FuelEquipmentProfile[]>(client.get(`/projects/${projectId}/fuel-management/equipment-profiles`)),
  createStorage: (projectId: string, body: Record<string, unknown>) => data<FuelStorage>(client.post(`/projects/${projectId}/fuel-management/storage`, body)),
  orders: (projectId: string, mine = false) => data<FuelOrder[]>(client.get(`/projects/${projectId}/fuel-management/orders`, { params: { mine } })),
  order: (orderId: string) => data<FuelOrder>(client.get(`/fuel-management/orders/${orderId}`)),
  createOrder: (projectId: string, body: Record<string, unknown>) => data<FuelOrder>(client.post(`/projects/${projectId}/fuel-management/orders`, body)),
  submitRequest: (projectId: string, body: Record<string, unknown>) => data<FuelOrder>(client.post(`/projects/${projectId}/fuel-management/requests`, body)),
  transitionOrder: (orderId: string, action: "submit" | "approve" | "reject" | "mark-ordered" | "cancel" | "close", body: Record<string, unknown> = {}) => data<FuelOrder>(client.post(`/fuel-management/orders/${orderId}/${action}`, body)),
  deliveries: (projectId: string) => data<FuelDelivery[]>(client.get(`/projects/${projectId}/fuel-management/deliveries`)),
  pendingConfirmations: (projectId: string) => data<FuelPendingConfirmation[]>(client.get(`/projects/${projectId}/fuel-management/deliveries/pending-confirmation`)),
  receiveFromProcurement: (projectId: string, body: Record<string, unknown>) => data<FuelDelivery>(client.post(`/projects/${projectId}/fuel-management/deliveries/from-procurement`, body)),
  recordDelivery: (orderId: string, body: Record<string, unknown>) => data<FuelDelivery>(client.post(`/fuel-management/orders/${orderId}/deliveries`, body)),
  verifyDelivery: (deliveryId: string, approve = true) => data<FuelDelivery>(client.post(`/fuel-management/deliveries/${deliveryId}/${approve ? "verify" : "reject"}`)),
  issues: (projectId: string, params?: { vehicle_id?: string; equipment_reference?: string; anomaly_only?: boolean }) =>
    data<FuelIssue[]>(client.get(`/projects/${projectId}/fuel-management/issues`, { params })),
  createIssue: (projectId: string, body: Record<string, unknown>) => data<FuelIssue>(client.post(`/projects/${projectId}/fuel-management/issues`, body)),
  createIssueWithEvidence: (projectId: string, body: Record<string, unknown>, files: Record<string, File | null>, onProgress?: (pct: number) => void) => {
    const form = new FormData(); form.append("payload", JSON.stringify(body));
    Object.entries(files).forEach(([key, file]) => { if (file) form.append(key, file); });
    return data<FuelIssue>(client.post(`/projects/${projectId}/fuel-management/issues-with-evidence`, form, {
      onUploadProgress: (event) => onProgress?.(event.total ? Math.round(event.loaded * 100 / event.total) : 0),
    }));
  },
  reverseIssue: (issueId: string, reason: string) => data<FuelIssue>(client.post(`/fuel-management/issues/${issueId}/reverse`, { reason })),
  reconciliations: (projectId: string) => data<FuelReconciliation[]>(client.get(`/projects/${projectId}/fuel-management/reconciliations`)),
  reconcile: (projectId: string, body: Record<string, unknown>) => data<FuelReconciliation>(client.post(`/projects/${projectId}/fuel-management/reconciliations`, body)),
  approveReconciliation: (id: string, reason: string) => data<FuelReconciliation>(client.post(`/fuel-management/reconciliations/${id}/approve`, { reason })),
  downloadReport: async (projectId: string, report: "orders" | "usage") => {
    const response = await client.get(`/projects/${projectId}/fuel-management/reports/${report}.csv`, { responseType: "blob" });
    const url = URL.createObjectURL(response.data); const anchor = document.createElement("a");
    anchor.href = url; anchor.download = `fuel-${report}.csv`; anchor.click(); URL.revokeObjectURL(url);
  },
};
